rfcs/0001-meta-schedule-autotensorir.md

```diff
@@ -0,0 +1,444 @@
+        ...
+
+    def apply(sch: Schedule, block: BlockRV) -> Union[Schedule, List[Schedule]]:
+        # write any python code, including:
+        # - analyze `block`
+        # - invoke schedule primitives in `sch`
+        # - do debug printing
+        ...
+```
+
+Method 2. A decorator as the syntactic sugar if the `initialize` method is empty, which converts the
+function to the `apply` method.
+
+```python
+@tir.as_composite_schedule(name="multi-level-tiling")
+def multi_level_tiling(sch: Schedule, block: BlockRV) -> Union[Schedule, List[Schedule]]:
+    # write any python code, including:
+    # - analyze `block`
+    # - invoke schedule primitives in `sch`
+    # - do debug printing
+    ...
+```
+
+**Customize exploration strategies in python**. Developers can implement any search algorithm in
+python as well by deriving from `PySearchPolicy`, and the syntax is identical to customizing with
+`PyCompositeSchedule`.
+
+**Other customizable components**. This list includes:
+
+* Cost model
+* Database
+* Measure callbacks
+* Feature extractor
+* Program builder & runner
+* Analysis methods
+* ...
+
+In a short summary, almost every component of the system is decoupled with each other and extensions
+could be easily plugged in.
+
+## 5. Drawbacks
+
+We are not aware of any drawbacks of the proposed system.
+
+## 6. Rationale and alternatives
+
+The system is designed with the principle of minimalism: different from alternative solutions, we do
+not require any change in existing codebase, or extra APIs to learn. It could potentially lower the
+bar of using automation systems. 
+
+Unifying manual scheduling, AutoTVM's semi automatic templates and AutoScheduler's (Ansor's) fully
+automatic sketch generation provides flexible way to balance injection new domain knowledge and
+automation.
+
+Flexibility in customization allows quick try-out on new tasks, new strategies and new hardware
+targets without deep knowledge of the system.
+
+## 7. Prior art
+
+**Tensor Expression (TE)** in TVM is a DSL that decouples compute and schedule, which provides
+convenient ways to handcraft optimized kernels for different hardware targets.
+
+**TensorIR** is the latest generation of TVM’s low-level IR. Its capability of eagerly applying
+schedule primitives opens the door for meta schedule, our proposed new-generation auto scheduling
+system.
+
+**AutoTVM** is the 1st generation automation framework in TVM, which requires developers to
+implement per-operator scheduling templates, and the system could handle the tuning process.
+
+**AutoScheduler (Ansor)** is the 2nd generation automation framework in TVM, whose built-in rules
+could automatically generate schedule templates for almost all the operators on CPU, GPU, etc.
+
+## 8. Unresolved questions
+
+**Supporting Control Flow and Assertions**
+
+Right now the meta schedule DSL does not support control flow. Although we didn’t see any real-world
```

Comment:
could you give an example of what control flow would look like in the DSL, even if not real-world?

Reply:
I am not super certain what the best syntax should look like. I added some explanation:

```markdown

**Control Flow**

The meta schedule DSL does not support control flow yet. Although there is no report of
real-world use case at the time of writing, it is possible that it could appear in some future
workloads. The best syntax of the control flow is not determined yet, but a working example could be
TensorFlow's `tf.cond`.

**Assertion**

Sampling instructions may lead to wrong schedules on CUDA, e.g. the resulting program uses too much
shared memory, too many threads, etc. It is detected by a postprocessor. To accelerate the process,
it is possible that we introduce an assertion statement that exits early if the GPU code is not
valid, and its syntax can be:

``"python
sch.assert(j_2 * j_2 <= 1024)
``"

```